File: app/src/main/java/com/chelmodeev/altimeter/logic/Advisor.kt
package com.chelmodeev.altimeter.logic

import com.chelmodeev.altimeter.model.Advice
import com.chelmodeev.altimeter.model.AdviceKind
import com.chelmodeev.altimeter.model.AdviceSeverity
import java.util.Locale
import kotlin.math.abs

data class AdvisorInput(
    val nowMs: Long,
    val altitude: Double?,
    val verticalSpeedMpm: Double?,
    val hasFix: Boolean,
    val locationGranted: Boolean,
    val spo2: Double?,
    val spo2AtMs: Long?,
    val heartRate: Long?,
    val heartRateAtMs: Long?,
    /** Тренд приведённого к уровню моря давления, гПа/ч; null — мало данных. */
    val pressureTrendHpaPerHour: Double?,
)

/**
 * Нейтральная полевая информация: вода, тренд погоды и качество GPS.
 */
class Advisor {

    fun evaluate(inp: AdvisorInput): List<Advice> {
        val out = mutableListOf<Advice>()

        inp.pressureTrendHpaPerHour?.let { t ->
            val v = String.format(Locale.getDefault(), "%.1f", abs(t))
            when {
                t <= -1.6 -> out += Advice(AdviceKind.PRESSURE_FALLING_FAST, AdviceSeverity.WARNING, v)
                t <= -0.8 -> out += Advice(AdviceKind.PRESSURE_FALLING, AdviceSeverity.CAUTION, v)
                t >= 1.2 -> out += Advice(AdviceKind.PRESSURE_RISING, AdviceSeverity.INFO)
                else -> Unit
            }
        }

        val alt = inp.altitude
        if (alt != null && alt >= 1500) {
            out += Advice(AdviceKind.HYDRATION, AdviceSeverity.INFO)
        }

        if (inp.locationGranted && !inp.hasFix) {
            out += Advice(AdviceKind.GPS_WEAK, AdviceSeverity.INFO)
        }

        return out
            .sortedByDescending { it.severity.ordinal }
            .take(4)
    }

}

File: ios/Altimeter/Engine/Advisor.swift
import Foundation

struct AdvisorInput {
    let now: Date
    let altitude: Double?
    let verticalSpeedMetersPerMinute: Double?
    let hasFix: Bool
    let locationAuthorized: Bool
    let oxygenPercent: Double?
    let oxygenDate: Date?
    let heartRate: Double?
    let heartRateDate: Date?
    let pressureTrendHPAperHour: Double?
}

struct Advisor {
    func evaluate(_ input: AdvisorInput) -> [Advice] {
        var result: [Advice] = []

        if let trend = input.pressureTrendHPAperHour {
            let value = String(format: "%.1f", abs(trend))
            if trend <= -1.6 {
                result.append(Advice(kind: .pressureFallingFast, severity: .warning, value: value))
            } else if trend <= -0.8 {
                result.append(Advice(kind: .pressureFalling, severity: .caution, value: value))
            } else if trend >= 1.2 {
                result.append(Advice(kind: .pressureRising, severity: .info, value: nil))
            }
        }

        if let altitude = input.altitude {
            if altitude >= 1_500 {
                result.append(Advice(kind: .hydration, severity: .info, value: nil))
            }
        }

        if input.locationAuthorized && !input.hasFix {
            result.append(Advice(kind: .gpsWeak, severity: .info, value: nil))
        }

        return result.sorted { $0.severity > $1.severity }.prefix(4).map { $0 }
    }
}
